### ISSUE TITLE
Document entities and correct use

### ISSUE THREAD

**Issue by aydun** [civicrm/civicrm-dev-docs#659]
The background for this is the discussion on https://lab.civicrm.org/dev/membership/issues/13 but applies more generally.

We have good docs for general usage, and for some aspects of development but we don't have anything that documents how the internal data structures relate and how to use them correctly (kinda ER but for humans!).

For example, if a membership with recurring contributions is created, what entities should result? eg membership, recurring contribution, contribution, etc.  When a payment is made on that recurring contribution what should result? - contributions, payment, membership payment, line item, financial transaction, updates to membership ...

The second part of that is how each of those things is created.  [The line Eileen pointed out](https://lab.civicrm.org/dev/membership/issues/13#note_21349) shows creating a lineitem will cause a membership payment to be created - unless it already exists, suggesting some ambiguity around whether a membership payment should have already been created. The documentation I have in mind would be something like: To create a membership with recurring contributions:  create Entity1 ..., Entity2 ..., Entity3 ... and then Entity4, Entity5, Entity6 will also be created.

Our docs and API tell us for example how to create a Membership Payment, but not the bigger picture of how to correctly record a payment for membership.

A clear picture of these entity relationships and processes would provide a more robust basis for refactoring, enhancements and extensions.

Thoughts?

**Comment by eileenmcnaughton** [comment]
@JoeMurray I believe there is still a doc on the old confluence somewhere....

@aydun specifically on memberships for any contribution that pays a membership it should be linked in 2 redundant ways

1) there should be a line item which has
entity_id = membership id
entity_table = civicrm_membership
contribution_id = contribution id
membership_type id = membership

It would be ideal if this was all that was needed but when we last had this discussion (years ago) it was still the case that some sites had contributions with incorrect line items and the membership payment records were not fully obsolete - as they ideally would be). Note the code to create a membership payment does attempt to link the line item too.

**Comment by davejenx** [comment]
@aydun Agreed that it's good to document this and thanks for getting it underway!

If Civi ever gets a major rewrite, I would like to see all this stuff much more abstracted and encapsulated. It seems to me that code in a form method or IPN callback not only shouldn't have to know about how to create & update the relevant related entities, it shouldn't be able to meddle at that level. The knowledge about how to do this should be encapsulated within a Membership object.

_"If to do were as easy as to know what were good to do, chapels had been churches, and poor men's cottages princes' palaces..."_

**Comment by eileenmcnaughton** [comment]
This is the documentation for the financial entites - https://wiki.civicrm.org/confluence/display/CRM/CiviAccounts+Data+Flow we really need to get this across

**Comment by MikeyMJCO** [comment]
I’ll pick that up next week!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

MICHAEL O'TOOLE
Managing Director
 T 029 2009 7097
 M 07769 901541
 E mikey@mjco.uk
WWW.MJCO.UK
54 Florentia Street
Cardiff, CF24 4PF
This email was sent whilst on-the-go! Please excuse any unintentional errors or brevity!
 We don't expect you to read a page of terms and conditions all we ask is that you let us know if we sent this to the wrong person and we'd like you to know that nothing in this email is intended to form a binding legal contract. 
 
 MJCO LTD is a company registered in Wales. Registered Number: 08483853. ​Registered Office: 54 Florentia Street, Cardiff CF24 4PF.
